Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp

```diff
@@ -0,0 +1,258 @@
+ExceptionOr<String> canonicalizeHost(String&& value, BaseURLStringType valueType)
+{
+    if (value.isEmpty())
+        return value;
+
+    if (valueType == BaseURLStringType::Pattern)
+        return value;
+
+    URL dummyURL("https://www.webkit.org"_s);
+    dummyURL.setHost(value);
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError };
+
+    return value;
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-an-ipv6-hostname
+ExceptionOr<String> canonicalizeIPv6Host(String&& value, BaseURLStringType valueType)
+{
+    if (valueType == BaseURLStringType::Pattern)
+        return value;
+
+    StringBuilder result;
+    result.reserveCapacity(result.length());
+
+    for (auto codepoint : StringView(value).codePoints()) {
+        if (isInvalidIPv6HostCodePoint(codepoint))
+            return Exception { ExceptionCode::TypeError };
+
+        result.append(codepoint);
+    }
+
+    return result.toString();
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-port, combined with https://urlpattern.spec.whatwg.org/#process-port-for-init
+ExceptionOr<String> canonicalizePort(String&& portValue, std::optional<StringView> protocolValue, BaseURLStringType portValueType)
+{
+    if (portValue.isEmpty())
+        return portValue;
+
+    if (portValueType == BaseURLStringType::Pattern)
+        return portValue;
+
+    URL dummyURL("https://www.webkit.org"_s);
+
+    if (protocolValue.has_value())
+        dummyURL.setProtocol(protocolValue.value());
+
+    dummyURL.setPort(parseInteger<uint16_t>(portValue));
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError };
+
+    int portAsInteger = static_cast<int>(dummyURL.port().value());
+    return String::number(portAsInteger);
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-an-opaque-pathname
+static ExceptionOr<String> canonicalizeOpaquePath(String&& value)
+{
+    URL dummyURL("https://www.webkit.org"_s);
+    dummyURL.setPath(value);
+
+    if (!dummyURL.isValid())
+        return Exception { ExceptionCode::TypeError };
+
+    return value;
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-pathname, combined with https://urlpattern.spec.whatwg.org/#process-pathname-for-init
+ExceptionOr<String> canonicalizePath(String&& pathnameValue, StringView protocolValue, BaseURLStringType pathnameValueType)
+{
+    if (pathnameValue.isEmpty())
+        return pathnameValue;
+
+    if (pathnameValueType == BaseURLStringType::Pattern)
+        return pathnameValue;
+
+    if (protocolValue == "ftp"_s || protocolValue == "file"_s
+        || protocolValue == "http"_s || protocolValue == "https"_s
+        || protocolValue == "ws"_s || protocolValue == "wss"_s) {
+
+        bool hasLeadingSlash = pathnameValue.startsWith("/"_s);
+        String modifiedValue = hasLeadingSlash ? ""_s : "/-"_s;
```

Comment:
`auto`. Note that this will resolve as an ASCIILiteral instead of constructing a String, which will be more efficient here.